File: src/day 4/solutions.py
from collections import defaultdict
from pyaoc.pyaoc import Env, Input

e = Env(4)
e.T("""MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX""", 18, 9)

xmas_str = "XMAS"
mas_str = "MAS"

def make_input_into_grid(file_input: Input):
    grid = defaultdict(str) | {(row_index, col_index): col for row_index, row in enumerate(file_input.get_valid_lines())
                               for col_index, col in enumerate(row)}
    return grid

def part1(file_input: Input):
    grid = make_input_into_grid(file_input)
    grid_positions = list(grid.keys())
    directions = -1,0,1
    total_sum = 0
    for row_direction in directions:
        for col_direction in directions:
            for row, col in grid_positions:
                word = [grid[row + row_direction * index, col + col_direction * index] for index in range(len(xmas_str))]
                total_sum += word == list(xmas_str)
    return total_sum

def part2(file_input: Input):
    grid = make_input_into_grid(file_input)
    grid_positions = list(grid.keys())
    directions = -1,0,1
    total_sum = 0
    search_str = list(mas_str), list(reversed(mas_str))
    for row, col in grid_positions:
        if row == 0 or col == 0: continue
        left_diagonal = [grid[row + direction, col + direction] for direction in directions]
        right_diagonal = [grid[row + direction, col - direction] for direction in directions]
        total_sum += (left_diagonal in search_str) and (right_diagonal in search_str)
    return total_sum

e.run_tests(1, part1)
e.run_main(1, part1)

e.run_tests(2, part2)
e.run_main(2, part2)

File: src/day 4/solutions.ts
import "@total-typescript/ts-reset";
import { showResult } from "../utils";
const defaultInput = await Bun.file("input.txt").text();

type CheckWordParams = {
	grid: string[];
	row: number;
	col: number;
	rowDirection: number;
	colDirection: number;
};

const directions = [-1, 0, 1];

const checkWord = ({
	grid,
	row,
	col,
	rowDirection,
	colDirection,
}: CheckWordParams): boolean => {
	const xmasStr = "XMAS";
	const wordLength = xmasStr.length;
	const lastRow = row + (wordLength - 1) * rowDirection;
	const lastCol = col + (wordLength - 1) * colDirection;
	if (
		lastRow < 0 ||
		lastCol < 0 ||
		lastRow >= grid.length ||
		lastCol >= grid[0].length
	) {
		return false;
	}
	const stringToCheck = Array(wordLength)
		.fill(0)
		.reduce((acc, _, i) => {
			return acc + grid[row + i * rowDirection][col + i * colDirection];
		}, "");
	return xmasStr === stringToCheck;
};

const count_xmas_in_directions = (
	grid: string[],
	row: number,
	col: number,
): number => {
	let count = 0;
	directions.forEach((rowDirection) => {
		directions.forEach((colDirection) => {
			if (
				checkWord({
					grid,
					row,
					col,
					rowDirection,
					colDirection,
				})
			) {
				count += 1;
			}
		});
	});
	return count;
};

const count_xmas = (lines: string[]): number => {
	let count = 0;
	for (let row = 0; row < lines.length; ++row) {
		for (let col = 0; col < lines[row].length; ++col) {
			if (lines[row][col] === "X") {
				count += count_xmas_in_directions(lines, row, col);
			}
		}
	}
	return count;
};

export const part1 = (input: string = defaultInput): number => {
	const res = count_xmas(input.split("\n"));
	return res;
};
const masStr = "MAS";

export const part2 = (input: string = defaultInput): number => {
	const searchStr = [masStr, [...masStr].reverse().join("")];
	const grid = input.split("\n");
	let count = 0;
	for (let row = 0; row < grid.length - 1; ++row) {
		if (row === 0) continue;
		for (let col = 0; col < grid[row].length - 1; ++col) {
			let leftDiagonal = "";
			let rightDiagonal = "";
			for (const direction of directions) {
				if (col === 0) continue;
				leftDiagonal = leftDiagonal + grid[row + direction][col + direction];
				rightDiagonal = rightDiagonal + grid[row + direction][col - direction];
			}
			if (
				searchStr.includes(leftDiagonal) &&
				searchStr.includes(rightDiagonal)
			) {
				count++;
			}
		}
	}
	return count;
};

showResult(part1(), 1);
showResult(part2(), 2);
